src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -332,14 +332,24 @@ private static int VerifyClientCertificate(int preverify_ok, IntPtr x509_ctx_ptr
 
         private static unsafe int AlpnServerSelectCallback(IntPtr ssl, out IntPtr outp, out byte outlen, IntPtr inp, uint inlen, IntPtr arg)
         {
+            outp = IntPtr.Zero;
+            outlen = 0;
+
+            if (arg == IntPtr.Zero)
+            {
+                return Interop.Ssl.SSL_TLSEXT_ERR_NOACK;
+            }
+
             GCHandle protocols = GCHandle.FromIntPtr(arg);
+            if (!protocols.IsAllocated || protocols.Target == null)
+            {
+                return Interop.Ssl.SSL_TLSEXT_ERR_NOACK;
+            }
+
             byte[] server = (byte[])protocols.Target;
 
-            fixed (byte* sp = server)
-            {
-                return Interop.Ssl.SslSelectNextProto(out outp, out outlen, (IntPtr)sp, (uint)server.Length, inp, inlen) == Interop.Ssl.OPENSSL_NPN_NEGOTIATED ?
+            return Interop.Ssl.SslSelectNextProto(out outp, out outlen, (IntPtr)protocols.AddrOfPinnedObject(), (uint)server.Length, inp, inlen) == Interop.Ssl.OPENSSL_NPN_NEGOTIATED ?
```

Comment:
Is the issue then that SslSelectNextProto uses the passed in buffer beyond the end of the synchronous call to the method?

Reply:
Yes exactly that ^^^^

Reply:
Actually not exactly that... its that you pass it back out with the "out IntPtr"

Reply:
I see.  The docs say:
```
The out value will point into either server or client, so it should be copied immediately.
```
This change ensures the server input is appropriately immovable.  Is that already true for the client input?

Reply:
Yes because the client input comes from OpenSSL this method is only called on the server side. As we don't store this but retrieve the value later from OpenSSL itself the lifetime of the client buffer is fine.